kernel_gateway/proposal.md

```diff
@@ -0,0 +1,90 @@
+
+Who is the Steering Council Advocate for this Subproject?
+
+* Kyle Kelley (`@rgbkrk`)
+
+## Subproject goals, scope and functionality
+
+### Problem
+
+Applications that use Jupyter kernels as execution engines outside of the traditional notebook / console user experience have started appearing on the web (e.g., [pyxie](https://github.com/oreillymedia/pyxie-static), [pyxie kernel server](https://github.com/oreillymedia/ipython-kernel), [Thebe](https://github.com/oreillymedia/thebe), [gist exec](https://github.com/rgbkrk/gistexec), [notebooks-to-dashboards](http://blog.ibmjstart.net/2015/08/22/dynamic-dashboards-from-jupyter-notebooks/), ). Today, these projects rely on:
+
+1. A _client_ (e.g., Thebe) that includes JavaScript code from Jupyter Notebook to request and communicate with kernels
+
+2. A _spawner_ (e.g., tmpnb) that provisions gateway servers to handle client kernel requests
+
+3. A _gateway_ (e.g., the entirety of Jupyter Notebook) that accepts [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) requests for kernels, isolates kernel workspaces (e.g., via Docker), and proxies web-friendly protocols (e.g., Websocket) to the kernel protocol (0mq)
+
+Maturing the Jupyter stack so that these efforts (and future ones) can find robust, common ground will require improvements in each of the above areas. Work is already underway to define a JavaScript library for communication with kernels and kernel provisioners (i.e., [jupyter/jupyter-js-services](https://github.com/jupyter/jupyter-js-services)). Discussion has started about defining an API for provisioning notebook servers as well ([binder project/binder#8](https://github.com/binder-project/binder/issues/8)), a topic that touches on the spawner concept above. This proposal focuses on the third area, the concept of a standard kernel provisioning and communication API.
+
+### Goals
+
+The goal of this incubator project is to **prototype and evaluate** possible solutions that satisfy the rising, novel kernel uses above, particularly with regard to the driving use cases documented below. The design and code of this incubator may become new Jupyter components, or folded into other relevant efforts underway, or discarded if better options arise as the evaluation proceeds.  At present, we do not know the **correct** design and implementation, but we believe there is value in trying the following:
+
+1. Using jupyter_client, jupyter_core, and pieces of jupyter/notebook (e.g., MappingKernelManager, etc.) to construct a headless kernel gateway that can talk to a cluster manager (e.g., Mesos).
+2. Implementing a websocket to 0mq bridge that can be placed in any Docker container that already runs a kernel, to allow web-friendly access to that kernel.
+3. Adding a new jupyter_client.WebsocketKernelManager that can be plugged into Jupyter Notebook or consumed by other tools to talk to kernels frontend by a websocket to 0mq bridge. (See use case #3 below).
```

Comment:
OK, cool. I think I understand now, but let me check. For the scenario where the actor slots are filled as follows:

1. client = Jupyter notebook frontend JS
2. server = the Jupyter notebook server, a Python/tornado webserver application
3. kernel-provider =  a remote service for launching kernels

The flow would be something like:

1. User opens a notebook document
2. Frontend for the notebook loads, including the client
3. Client starts a new session
4. Server requests a new kernel from kernel-provider on the client's behalf (REST, not websocket)
5. Client connects directly to the websocket associated with the session (cross domain)

versus the client talking to the server in step #5 and the server having to proxy the websocket to the kernel gateway.

Is this right?